src/main/java/org/apache/sling/xss/impl/CustomPolicy.java

```diff
@@ -0,0 +1,265 @@
+import org.apache.sling.xss.impl.xml.Policy;
+import org.apache.sling.xss.impl.xml.Tag;
+import org.owasp.html.AttributePolicy;
+import org.owasp.html.HtmlPolicyBuilder;
+import org.owasp.html.PolicyFactory;
+
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableSet;
+
+public class CustomPolicy {
+    private PolicyFactory policyFactory;
+    private List<String> onInvalidRemoveTagList = new ArrayList<>();
+    private Map<String, AttributePolicy> dynamicAttributesPolicyMap = new HashMap<>();
+    private CssValidator cssValidator;
+
+    public CustomPolicy(Policy policy) {
+        removeAttributeGuards();
+        HtmlPolicyBuilder policyBuilder = new HtmlPolicyBuilder();
+
+        cssValidator = new CssValidator(policy.getCssPolicy());
+
+        // ------------ this is for the global attributes -------------
+        Map<String, Attribute> globalAttributes = policy.getGlobalAttributes();
+        for (Attribute attribute : globalAttributes.values()) {
+
+            if (attribute.getOnInvalid().equals("removeTag")) {
+                onInvalidRemoveTagList.add(attribute.getName());
+            }
+
+            if (CssValidator.STYLE_ATTRIBUTE_NAME.equals(attribute.getName())) {
+                // we match style tags separately
+                policyBuilder.allowAttributes(attribute.getName()).matching(cssValidator.newCssAttributePolicy())
+                        .globally();
+            } else {
+                List<String> allowedValuesFromAttribute = attribute.getLiterals();
+                if (allowedValuesFromAttribute != null && allowedValuesFromAttribute.size() > 0) {
+                    for (String allowedValue : allowedValuesFromAttribute) {
+                        policyBuilder.allowAttributes(attribute.getName()).matching(true, allowedValue).globally();
+                    }
+
+                }
+                List<Pattern> regexsFromAttribute = attribute.getPatternList();
+                if (regexsFromAttribute != null && regexsFromAttribute.size() > 0) {
+                    policyBuilder.allowAttributes(attribute.getName()).matching(matchesToPatterns(regexsFromAttribute))
+                            .globally();
+                } else {
+                    policyBuilder.allowAttributes(attribute.getName()).globally();
+                }
+
+            }
+        }
+
+        // ------------ this is for the allowed emty tags -------------
+        List<String> allowedEmptyTags = policy.getAllowedEmptyTags();
+        for (String allowedEmptyTag : allowedEmptyTags) {
+            policyBuilder.allowWithoutAttributes(allowedEmptyTag);
+        }
+
+        // ------------ this is for the tag rules -------------
+        Map<String, Tag> tagMap = policy.getTagRules();
+        for (Map.Entry<String, Tag> tag : tagMap.entrySet()) {
+
+            String tagAction = tag.getValue().getAction();
+            switch (tagAction) {
+            // Tag.action
+            case "truncate":
```

Comment:
We should extract these string values as constants somewhere and reference them from the code.

Reply:
Resolved: 1eebd8227f8ec90fef211eaad9cbcdb7ded10c03